File: Tests/LedgelingsCoreTests/SparksTests.swift
import CoreGraphics
import Testing
@testable import LedgelingsCore

/// The little burst of pixel stars when two creatures bump.
@Suite struct SparksTests {
    @Test func aBurstThrowsStarsUpAndTheyFallBackAndDie() {
        var rng = SeededRNG(state: 7)
        var sparks = Sparks()
        sparks.burst(at: CGPoint(x: 100, y: 10), inward: CGVector(dx: 0, dy: 1), count: 8, using: &rng)
        #expect(sparks.alive.count == 8)
        sparks.update(dt: 0.1)
        #expect(sparks.alive.allSatisfy { $0.position.y > 10 }, "all thrown into the screen at first")
        for _ in 0..<20 { sparks.update(dt: 0.05) }
        #expect(sparks.alive.isEmpty, "gone after about a second")
    }

    @Test func starsFadeAsTheyAge() {
        var rng = SeededRNG(state: 8)
        var sparks = Sparks(life: 1)
        sparks.burst(at: .zero, inward: CGVector(dx: 0, dy: 1), count: 3, using: &rng)
        sparks.update(dt: 0.2)
        let early = sparks.alive.map(\.opacity)
        sparks.update(dt: 0.6)
        let late = sparks.alive.map(\.opacity)
        #expect(zip(early, late).allSatisfy { $0 > $1 })
        #expect(late.allSatisfy { $0 > 0 && $0 < 0.5 })
    }

    @Test func gravityPullsThemBackTowardTheEdgeTheyCameFrom() {
        var rng = SeededRNG(state: 9)
        var sparks = Sparks(life: 5)
        sparks.burst(at: .zero, inward: CGVector(dx: -1, dy: 0), count: 4, using: &rng)   // a right-hand wall: "up" is leftwards
        sparks.update(dt: 0.1)
        let before = sparks.alive.map(\.velocity.dx)
        sparks.update(dt: 0.5)
        let after = sparks.alive.map(\.velocity.dx)
        #expect(zip(before, after).allSatisfy { $1 > $0 }, "each star's leftward speed shrinks")
    }
}

File: win/LedgelingsCore.Tests/SparksTests.cs
namespace Ledgelings.Core.Tests;

/// <summary>The little burst of pixel stars when two creatures bump.</summary>
public class SparksTests
{
    [Fact]
    public void ABurstThrowsStarsUpAndTheyFallBackAndDie()
    {
        var rng = new Random(7);
        var sparks = new Sparks();
        sparks.Burst(new Pt(100, 10), new Vec(0, 1), 8, rng);
        Assert.Equal(8, sparks.Alive.Count);
        sparks.Update(0.1);
        Assert.True(sparks.Alive.All(s => s.Position.Y > 10), "all thrown into the screen at first");
        for (int i = 0; i < 20; i++) sparks.Update(0.05);
        Assert.True(sparks.Alive.Count == 0, "gone after about a second");
    }

    [Fact]
    public void StarsFadeAsTheyAge()
    {
        var rng = new Random(8);
        var sparks = new Sparks(life: 1);
        sparks.Burst(Pt.Zero, new Vec(0, 1), 3, rng);
        sparks.Update(0.2);
        var early = sparks.Alive.Select(s => s.Opacity).ToList();
        sparks.Update(0.6);
        var late = sparks.Alive.Select(s => s.Opacity).ToList();
        Assert.True(early.Zip(late).All(p => p.First > p.Second));
        Assert.True(late.All(o => o > 0 && o < 0.5));
    }

    [Fact]
    public void GravityPullsThemBackTowardTheEdgeTheyCameFrom()
    {
        var rng = new Random(9);
        var sparks = new Sparks(life: 5);
        sparks.Burst(Pt.Zero, new Vec(-1, 0), 4, rng);   // a right-hand wall: "up" is leftwards
        sparks.Update(0.1);
        var before = sparks.Alive.Select(s => s.Velocity.Dx).ToList();
        sparks.Update(0.5);
        var after = sparks.Alive.Select(s => s.Velocity.Dx).ToList();
        Assert.True(before.Zip(after).All(p => p.Second > p.First), "each star's leftward speed shrinks");
    }
}
